mne/tests/test_epochs.py

```diff
@@ -61,14 +60,162 @@ def _get_data():
 clean_warning_registry()  # really clean warning stack
 
 
+def test_reject():
+    """Test epochs rejection
+    """
+    raw, events, picks = _get_data()
+    # cull the list just to contain the relevant event
+    events = events[events[:, 2] == event_id, :]
+    selection = np.arange(3)
+    drop_log = [[]] * 3 + [['MEG 2443']] * 4
+    assert_raises(TypeError, pick_types, raw)
+    picks_meg = pick_types(raw.info, meg=True, eeg=False)
+    assert_raises(TypeError, Epochs, raw, events, event_id, tmin, tmax,
+                  picks=picks, preload=False, reject='foo')
+    assert_raises(ValueError, Epochs, raw, events, event_id, tmin, tmax,
+                  picks=picks_meg, preload=False, reject=dict(eeg=1.))
+    assert_raises(KeyError, Epochs, raw, events, event_id, tmin, tmax,
+                  picks=picks, preload=False, reject=dict(foo=1.))
+
+    data_7 = dict()
+    keep_idx = [0, 1, 2]
+    for preload in (True, False):
+        for proj in (True, False):
+            # no rejection
+            epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                            preload=preload)
+            assert_raises(ValueError, epochs.drop_bad_epochs, reject='foo')
+            epochs.drop_bad_epochs()
+            assert_equal(len(epochs), len(events))
+            assert_array_equal(epochs.selection, np.arange(len(events)))
+            assert_array_equal(epochs.drop_log, [[]] * 7)
+            if proj not in data_7:
+                data_7[proj] = epochs.get_data()
+            assert_array_equal(epochs.get_data(), data_7[proj])
+
+            # with rejection
+            epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                            reject=reject, preload=preload)
+            epochs.drop_bad_epochs()
+            assert_equal(len(epochs), len(events) - 4)
+            assert_array_equal(epochs.selection, selection)
+            assert_array_equal(epochs.drop_log, drop_log)
+            assert_array_equal(epochs.get_data(), data_7[proj][keep_idx])
+
+            # rejection post-hoc
+            epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                            preload=preload)
+            epochs.drop_bad_epochs()
+            assert_equal(len(epochs), len(events))
+            assert_array_equal(epochs.get_data(), data_7[proj])
+            epochs.drop_bad_epochs(reject)
+            assert_equal(len(epochs), len(events) - 4)
+            assert_equal(len(epochs), len(epochs.get_data()))
+            assert_array_equal(epochs.selection, selection)
+            assert_array_equal(epochs.drop_log, drop_log)
+            assert_array_equal(epochs.get_data(), data_7[proj][keep_idx])
+
+            # rejection twice
+            reject_part = dict(grad=1100e-12, mag=4e-12, eeg=80e-6, eog=150e-6)
```

Comment:
On Sun, Jun 21, 2015 at 1:51 AM, Eric Larson <notifications@github.com>
wrote:

> In mne/tests/test_epochs.py
> <https://github.com/mne-tools/mne-python/pull/2211#discussion_r32889151>:
>
> > +
> > +            # rejection post-hoc
> > +            epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
> > +                            preload=preload)
> > +            epochs.drop_bad_epochs()
> > +            assert_equal(len(epochs), len(events))
> > +            assert_array_equal(epochs.get_data(), data_7[proj])
> > +            epochs.drop_bad_epochs(reject)
> > +            assert_equal(len(epochs), len(events) - 4)
> > +            assert_equal(len(epochs), len(epochs.get_data()))
> > +            assert_array_equal(epochs.selection, selection)
> > +            assert_array_equal(epochs.drop_log, drop_log)
> > +            assert_array_equal(epochs.get_data(), data_7[proj][keep_idx])
> > +
> > +            # rejection twice
> > +            reject_part = dict(grad=1100e-12, mag=4e-12, eeg=80e-6, eog=150e-6)
>
> That is actually tested on the lines we are commenting on... reject_part
> is more lenient and only gets rid of one epoch, whereas reject gets rid
> of four of them.
>
> I guess the question is, how should this function? Currently, the
> functionality is that you can pass reject and flat as many times as you
> want to drop_bad_epochs, and it will drop any existing epochs that meet
> those criteria.
>
> An alternative would be to make it so that you can only do drop_bad_epochs
> with a non-None reject and flat exactly once. Is that what you want?
>
yeah I think this is exactly what @agramfort wanted. Any other opinions?
@teonlamont?

> That would allow you to create epochs with reject = flat = None with any
> state of proj and preload, and then at some point later apply any
> rejection criteria you want.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2211/files#r32889151>.
>